### PR TITLE
feat: ENS integration

### PR DESCRIPTION
**Pull request by tianxiemaochiyu**

### CODE REVIEW

File: src/errors/AllDIDError.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 export enum AllDIDErrorCode {
   DidIsNotSupported = 'DidIsNotSupported',
+  DidIsInvalid = 'DidIsInvalid',
```
Comment (zgayjjf) [inline src/errors/AllDIDError.ts:3]: Seems not in use

File: src/services/EnsService.ts [hunk 2]
```diff
@@ -23,68 +36,212 @@ function getEnsAddress (networkId: string) {
...
+
+function getAddrRecordKeys () {
+  return ['ETH', 'BTC', 'LTC', 'DOGE']
+}
+
+async function batchRequest (
+  reqArray: any[],
+  callback: (res: any) => void
+): Promise<any> {
+  const BatchRequest = new Web3().BatchRequest
```
Comment (zgayjjf) [inline src/services/EnsService.ts:84]: Can `ethers` replace this?

File: package.json [hunk 3]
```diff
@@ -57,9 +56,14 @@
     "vite": "^2.9.14"
   },
   "dependencies": {
+    "@ensdomains/ens": "^0.6.2",
```
Comment (zgayjjf) [inline package.json:59]: Seems not in use.

File: src/services/EnsService.ts [hunk 4]
```diff
@@ -23,68 +36,212 @@ function getEnsAddress (networkId: string) {
...
-  addrs (name: string, keys?: string | string[]): Promise<RecordItemAddr[]> {
-    return null
-  }
-
-  addr (name: string): Promise<RecordItemAddr> {
-    return null
+    if (!this.isSupported(name)) return
+    const nameArray = name.split('.')
+    const label = nameArray[nameArray.length - 1]
+    const tokenID: string = new BigNumber(labelhash(label)).toString(16)
```
Comment (zgayjjf) [inline src/services/EnsService.ts:152]: `labelhash` is already hex, what's the purpose of BigNumber?

**Review by zgayjjf — CHANGES_REQUESTED** [review]

**Review by zgayjjf — CHANGES_REQUESTED** [review]

File: tests/services/ENSService.test.ts [hunk 5]
```diff
@@ -9,8 +9,92 @@ describe('ENSService', () => {
       expect(ensService.isSupported('abc.eth')).toBe(true)
     })
 
-    it('do not support abc.bit', () => {
-      expect(ensService.isSupported('abc.bit')).toBe(false)
+    it('do not support abc.bit', async () => {
+      await expect(ensService.isSupported('abc.bit')).toBe(false)
```
Comment (zgayjjf) [inline tests/services/ENSService.test.ts:13]: Is `await` necessary?

File: tests/services/ENSService.test.ts [hunk 6]
```diff
@@ -9,8 +9,92 @@ describe('ENSService', () => {
...
+    })
+
+    it('address that owner of leon.eth is valid', async () => {
+      const owner = await ensService.owner('leon.eth')
+      expect(owner).toMatch(/^0x[0-9A-Fa-f]{40}$/)
+    })
+
+    it('tokenId that owner of leon.eth is valid', async () => {
+      const tokenId = await ensService.tokenId('leon.eth')
+      expect(tokenId).toMatch(/^0x[0-9A-Fa-f]{64}$/)
```
Comment (zgayjjf) [inline tests/services/ENSService.test.ts:33]: `TokenId` of a specific name if fixed. So the test result should be a plain string.

File: src/services/EnsService.ts [hunk 7]
```diff
@@ -23,68 +29,183 @@ function getEnsAddress (networkId: string) {
...
+      return [
+        {
+          ...record,
+          symbol: keys.toUpperCase(),
+        },
+      ]
+    }
+  }
+
+  async addr (name: string): Promise<RecordItemAddr> {
```
Comment (zgayjjf) [inline src/services/EnsService.ts:179]: The interface should be `addr(name: string, key: string)`.
Users can either pass in `coin_type` or `symbol` of the token.

File: tests/services/ENSService.test.ts [hunk 8]
```diff
@@ -9,8 +9,92 @@ describe('ENSService', () => {
...
+          symbol: 'ETH',
+          ttl: 0,
+          type: 'address',
+          value: '0xC72B6f66017246d6A7f159F5C2BF358188AD9ECa'
+        }
+      ])
+    })
+
+    it('leont.eth\'s reverse is invalid', async () => {
+      const name = await ensService.addrs('0xC72B6f66017246d6A7f159F5C2BF358188AD9ECa')
```
Comment (zgayjjf) [inline tests/services/ENSService.test.ts:91]: Why is that invalid?

**Review by zgayjjf — CHANGES_REQUESTED** [review]

File: .eslintrc.js [hunk 9]
```diff
@@ -1,3 +1,3 @@
 module.exports = {
-  extends: 'blockabc/typescript',
+  extends: 'plugin:blockabc/typescript',
```
[the remark this answers is not in the document]
Comment (tianxiemaochiyu) [reply .eslintrc.js:2]: Solve the problem that the `blockabc/typescript` shared configuration does not work under the VSCode editor

File: package.json [hunk 10]
```diff
@@ -57,9 +57,12 @@
     "vite": "^2.9.14"
   },
   "dependencies": {
+    "@ensdomains/address-encoder": "^0.1.7",
+    "@ensdomains/ens-contracts": "^0.0.15",
```
Comment (zgayjjf) [inline package.json:61]: What's the purpose of `@ensdomains/ens-contracts`?

File: tests/services/ENSService.test.ts [hunk 11]
```diff
@@ -9,8 +9,97 @@ describe('ENSService', () => {
...
+
+    
+
+    it('leont.eth\'s registryAddress is valid', async () => {
+      const name = await ensService.addrs('leont.eth')
+      expect(name).toEqual('0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e')
+    })
+  })
+
+  describe('isSupported', () => {
```
Comment (zgayjjf) [inline tests/services/ENSService.test.ts:98]: Wrong test name

**Review by zgayjjf — CHANGES_REQUESTED** [review]

File: src/services/EnsService.ts [hunk 12]
```diff
@@ -23,68 +29,183 @@ function getEnsAddress (networkId: string) {
...
 
   isRegistered (name: string): Promise<boolean> {
-    return this.ens.exist(name)
+    return this.isSupported ? this.owner(name).then((owner) => !!owner) : null
   }
 
   isAvailable (name: string): Promise<boolean> {
-    return this.isRegistered(name).then(isRegistered => !isRegistered)
+    return this.isSupported
+      ? this.isRegistered(name).then((isRegistered) => !isRegistered)
```
[the remark this answers is not in the document]
Comment (zgayjjf) [reply src/services/EnsService.ts]: Where is Error thrown when unsupported?

File: src/services/EnsService.ts [hunk 13]
```diff
@@ -23,68 +31,188 @@ function getEnsAddress (networkId: string) {
...
+    'com.twitter',
+    'org.telegram',
+  ]
+}
+
+function getAddrRecordKeys (): string[] {
+  return ['ETH', 'BTC', 'LTC', 'DOGE']
+}
+
+function coinTypeOrcoinId2Symbol (coinType: string): string {
```
Comment (zgayjjf) [inline src/services/EnsService.ts:65]: What is `coinId`? There seems to be no such thing.
Even if such a thing exists, the function's name should be `coinTypeOrCoinId2Symbol`. Pay attention to the Uppercase

**Review by zgayjjf — CHANGES_REQUESTED** [review]

**Review by zgayjjf — APPROVED** [review]

**Pull request merged**